flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/FunctionCatalog.java

```diff
@@ -242,21 +242,17 @@ public void registerTempCatalogScalarFunction(ObjectIdentifier oi, ScalarFunctio
 			CatalogFunction catalogFunction = catalog.getFunction(
 				new ObjectPath(catalogManager.getCurrentDatabase(), functionName));
 
-			if (catalog.getTableFactory().isPresent() &&
-				catalog.getTableFactory().get() instanceof FunctionDefinitionFactory) {
-
-				FunctionDefinitionFactory factory = (FunctionDefinitionFactory) catalog.getTableFactory().get();
-
-				userCandidate = factory.createFunctionDefinition(functionName, catalogFunction);
-
-				return Optional.of(
-					new FunctionLookup.Result(
-						ObjectIdentifier.of(catalogManager.getCurrentCatalog(), catalogManager.getCurrentDatabase(), name),
-						userCandidate)
-				);
+			if (catalog.getFunctionDefinitionFactory().isPresent()) {
+				userCandidate = catalog.getFunctionDefinitionFactory().get().createFunctionDefinition(functionName, catalogFunction);
 			} else {
-				// TODO: should go through function definition discover service
+				userCandidate = FunctionDefinitionUtil.createFunctionDefinition(functionName, catalogFunction);
```

Comment:
I'm not sure if FunctionCatalog should make this call and it's probably better to let Catalog implementation do whatever needed.

We probably can make this call in GenericInMororyCatalog's FunctionDefinitionFactory.createFunctionDefinition().

What do you think?

Reply:
This has nothing to do with GenericInMemoryCatalog, thus I'm not so sure what you mean. Maybe you can elaborate more?

As we discussed this morning, catalog can provide its own function definition factory; if not, the function definition instantiation will go thru the default logic. Just like how table instantiation logic works

Reply:
My point is, if the catalog doesn't provide a FunctionDefFactory, we should not call FunctionDefinitionUtil.createFunctionDefinition() because we don't know if that's a catalog implementation wants. It's better for a catalog implementation to explicitly invoke this call, if that's what it needs.

Reply:
not sure if I understand it from an end-to-end perspective, can you provide some code example?

if the catalog doesn't provide a func def factory, what would planner do then? FunctionDefinitionUtil.createFunctionDefinition() serves as the default logic if users don't provide a func def factory to instantiate func def; if they want customized func def instantiation logic, they should provide a func def factory.

Reply:
uhn, I guess this is fine. Especially, the API is consistent.